### ISSUE TITLE
Clippy emitting wrong lint for unsafe

### ISSUE THREAD

**Issue by cristeigabriel** [rust-lang/rust-clippy#8052]
### Summary

```
unsafe fn no(arg: i32) -> i32 {
    arg
}

---

warning: this if-then-else expression returns a bool literal
  --> src\main.rs:6:5
   |
6  | /     if unsafe { no(4) } & 1 != 0 {
8  | |     } else {
9  | |         false
10 | |     }
   | |_____^ help: you can reduce it to: `unsafe { no(4) } & 1 != 0
```

The produced expression above is not valid, the `& 1 != 0` should be inside the `unsafe` block.

### Reproducer

I tried this code:

```rust
unsafe fn no(arg: i32) -> i32 {
    arg
}

fn state() -> bool {
    if unsafe { no(4) } & 1 != 0 {
        true
    } else {
        false
    }
}

fn main() {
    println!("{}", if state() { "hello" } else { "bye" });
}
```

I expected to see this happen:
Upon running `cargo clippy` I'd expect it to reduce state to `unsafe { no(4) & 1 != 0 }` 

Instead, this happened:
It reduced state to `unsafe { no(4) } & 1 != 0 ` which is invalid syntax

### Version

```text
rustc 1.59.0-nightly (686e313a9 2021-11-27)
binary: rustc
commit-hash: 686e313a9aa14107c8631ffe48fa09110a7692db
commit-date: 2021-11-27
host: i686-pc-windows-msvc
release: 1.59.0-nightly
LLVM version: 13.0.0
```

**Comment by cristeigabriel** [comment]
It seems that the Rust lexer thinks that I want to do a borrow, when I want to do the AND operation. both `&1 != 0` and `& 1 != 
0` outside of `unsafe` don't work. is this a lexer bug, or intended?

**Comment by llogiq** [comment]
The problem is that an `unsafe` block is still a block, so if I understand correctly it needs parentheses to be parsed as expression here.
